common/vm_get_primary_nic.yml

```diff
@@ -18,16 +18,16 @@
       - vm_guest_mac_ipv4 | length >= 1
     fail_msg: "Not found valid IPv4 address for VM '{{ vm_name }}'"
 
-- name: Display the VM MAC and IP address dict
+- name: "Display the VM MAC and IP address dict"
   ansible.builtin.debug: var=vm_guest_mac_ipv4
 
-- block:
-    - name: "Get IP of VM '{{ vm_name }}' primary NIC with MAC address {{ vm_primary_mac }}"
-      ansible.builtin.set_fact:
-        vm_primary_ip: "{{ vm_guest_mac_ipv4[vm_primary_mac] }}"
+- name: "Get IP of VM '{{ vm_name }}' primary NIC with MAC address {{ vm_primary_mac }}"
+  ansible.builtin.set_fact:
+    vm_primary_ip: "{{ vm_guest_mac_ipv4[vm_primary_mac] }}"
   when: vm_primary_mac is defined and vm_primary_mac in vm_guest_mac_ipv4.keys()
 
-- block:
+- name: "Test VM's IP connection to get VM's primary NIC"
+  block:
     - name: "Set the only network interface as VM primary NIC of '{{ vm_name }}'"
       ansible.builtin.set_fact:
         vm_primary_mac: "{{ (vm_guest_mac_ipv4.keys() | list)[0] }}"
```

Comment:
the block when condition is 
`
  when: >
     (vm_primary_mac is undefined) or
     (not vm_primary_mac) or
(vm_primary_mac not in vm_guest_mac_ipv4.keys())
`
after this setting, the following tasks will be executed?

Reply:
No. The following tasks are only applicable when VM has more than one IPv4 addresses. In this block, as long as vm_primary_mac is set, the following tasks can be skipped.

Reply:
Then the task in block can be put out of the block and changed since the set_fact variables are the when condition variables, which we need to avoid.
```
- name: "Set the only network interface as VM primary NIC of '{{ vm_name }}'"
   ansible.builtin.set_fact:
        vm_primary_mac: "{{ (vm_guest_mac_ipv4.keys() | list)[0] }}"
        vm_primary_ip: "{{ (vm_guest_mac_ipv4.values() | list)[0] }}"
   when: 
          - vm_guest_mac_ipv4 | length == 1
          - not vm_primary_ip
```